# **Inteligencia Artificial con Aplicaciones en Economía I**

- 👩‍🏫 **Profesora:** [Lina María Castro](https://www.linkedin.com/in/lina-maria-castro)  
- 📧 **Email:** [lmcastroco@gmail.com](mailto:lmcastroco@gmail.com)  
- 🎓 **Universidad:** Universidad Externado de Colombia - Facultad de Economía

# **Parte práctica parcial: Preparación y limpieza de datos**🧹

**IMPORTANTE**: Guarda una copia de este notebook en tu Google Drive o computador.

**Nombre estudiante:**

**Forma de entrega**

Jupyter Notebook publicado en su cuenta de Github con el nombre “Practica_limpieza_nombre_apellido_estudiante.ipynb”. Máximo a las 9 a.m., debes enviar link del notebook al correo lina.castro6@uexternado.edu.co, de lo contrario, no será tenido en cuenta.

**Instrucciones Generales**

- Puedes utilizar tus apuntes y notebooks y entrar al repositorio del curso. Sin embargo, **NO puedes utilizar IA para desarrollar tus respuestas**. **En caso de descubrir que estás usando IA** para el desarrollo de los ejercicios, **tu nota** de este segundo parcial **será cero**.
- Completa el código en las celdas marcadas con `### TU CÓDIGO AQUÍ ###`. Puedes añadir más celdas si lo requieres.
- Después de cada punto, ejecuta la celda de `assert` que se proporciona. Si no aparece ningún error, ¡tu respuesta es correcta!
- No modifiques las celdas de `assert` ni las celdas de Markdown.

**Objetivos de Aprendizaje**

1. **Aplicar técnicas de importación y consolidación de múltiples fuentes de datos** utilizando la librería `pandas`.
2.  **Ejecutar un proceso completo de limpieza de datos**, incluyendo el manejo de valores nulos, duplicados y la estandarización de tipos de datos.
3.  **Preparar un conjunto de datos real (GEIH del DANE)** para un análisis económico, sentando las bases para futuros modelos de machine learning.

# Ejercicio 1: Importar y unir datos (20%)

El DANE publica los microdatos de la GEIH mensualmente. Tu primera tarea es cargar los datos de los primeros seis meses de 2025 y armar una sola base.

**Los archivos necesarios los encuentras en el repositorio del curso, carpeta "datasets", carpeta "geih_2025".**

1. Usa un ciclo "for" para abrir los archivos CSV, ten en cuenta que estas bases usan como separador ";" y encoding='latin1'. La variable 'DPTO' y 'RAMA4D_R4' debes cargarlas como string.

In [192]:
import pandas as pd
import os

files = ['/content/df_ene.csv', '/content/df_feb.csv', '/content/df_mar.csv', '/content/df_abr.csv', '/content/df_may.csv',
         '/content/df_jun.csv']

for file in files:
    df = pd.read_csv(file,sep=';',encoding='latin1',dtype={'DPTO': str, 'RAMA4D_R4': str})

/tmp/ipython-input-1848172055.py:9: DtypeWarning: Columns (71,132,176,185,190,201) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file,sep=';',encoding='latin1',dtype={'DPTO': str, 'RAMA4D_R4': str})
/tmp/ipython-input-1848172055.py:9: DtypeWarning: Columns (71,132,176,188,190,201) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file,sep=';',encoding='latin1',dtype={'DPTO': str, 'RAMA4D_R4': str})
/tmp/ipython-input-1848172055.py:9: DtypeWarning: Columns (71,132,176,185,188,190,201) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file,sep=';',encoding='latin1',dtype={'DPTO': str, 'RAMA4D_R4': str})
/tmp/ipython-input-1848172055.py:9: DtypeWarning: Columns (71,132,176,188,190,201) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file,sep=';',encoding='latin1',dtype={'DPTO': str, 'RAMA4D_R4': str})
/tmp/ipython

2. Une todos los dataframes para obtener una base de la GEIH con toda la información de enero a junio de 2025. El nuevo dataframe debe llamarse `df_geih`

In [193]:
df_geih = pd.concat(dataframes, ignore_index=True)

3. Reinicia el índice de `df_geih`. Asegúrate de que el índice antiguo no se añada como una nueva columna.

In [194]:
df_geih = df_geih.reset_index(drop=True)

In [195]:
# Celda de verificación (no modificar)
assert df_geih.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_geih.shape[0]}."
assert df_geih.shape[1] == 245, f"Error en las columnas. Se esperaban 245, pero se encontraron {df_geih.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
assert 'index' not in df_geih.columns, "El índice antiguo no debe ser una columna."
assert df_geih.index.is_unique, "El índice debe ser único después de reiniciarlo."
print(f"¡Punto 1 completado con éxito! ✅ El DataFrame consolidado tiene {df_geih.shape[0]} filas y {df_geih.shape[1]} columnas.")

¡Las dimensiones del DataFrame son correctas! ✅
¡Punto 1 completado con éxito! ✅ El DataFrame consolidado tiene 408423 filas y 245 columnas.


# Ejercicio 2:  Seleccionar y renombrar columnas (5%)

La GEIH contiene cientos de variables. Para nuestro análisis del mercado laboral, nos enfocaremos en un subconjunto.

Adicionalmente, los nombres de las columnas en la GEIH son códigos. Para hacer el análisis más intuitivo, vamos a renombrarlas.

1. Crea un nuevo DataFrame llamado `df_empleo` que contenga únicamente las columnas que están en la lista `columnas_interes`.

In [196]:
columnas_interes = ['DIRECTORIO', 'SECUENCIA_P', 'ORDEN', 'MES', 'CLASE', 'DPTO', 'FEX_C18', 'PT', 'P3271', 'P6040', 'P6080', 'P3043', 'OCI', 'P6430', 'RAMA4D_R4', 'INGLABO']

In [197]:
df_empleo = df_geih[columnas_interes]

In [198]:
# Celda de verificación (no modificar)
assert len(df_empleo.columns) == 16, "El número de columnas es incorrecto."
assert all(col in df_empleo.columns for col in columnas_interes), "Faltan columnas o los nombres son incorrectos."
print("¡Selección de columnas completada con éxito! ✅")
print("Dimensiones del nuevo DataFrame:", df_empleo.shape)

¡Selección de columnas completada con éxito! ✅
Dimensiones del nuevo DataFrame: (408423, 16)


2. Renombra las columnas del DataFrame `df_empleo` utilizando el diccionario `mapa_nombres`.

In [199]:
mapa_nombres = {
'DIRECTORIO': 'id_vivienda',
'SECUENCIA_P': 'id_hogar',
'ORDEN': 'id_persona',
'MES': 'mes',
'CLASE': 'clase',
'DPTO': 'departamento',
'FEX_C18': 'factor_expansion',
'PT': 'poblacion_total',
'P3271': 'sexo',
'P6040': 'edad',
'P6080': 'grupo_etnico',
'P3043': 'nivel_educativo',
'OCI': 'ocupados',
'P6430': 'categoria_ocupacional',
'RAMA4D_R4': 'cod_ciiu',
'INGLABO': 'ingreso_laboral'}

In [200]:
df_empleo = df_empleo.rename(columns=mapa_nombres)

In [201]:
# Celda de verificación (no modificar)
nombres_nuevos_esperados = list(mapa_nombres.values())
condicion_exitosa = all(col in df_empleo.columns for col in nombres_nuevos_esperados)
assert condicion_exitosa, "No todas las columnas esperadas se encuentran en el DataFrame. Revisa el diccionario o la operación de renombrado."
print("¡Verificación exitosa! Todas las columnas fueron renombradas correctamente. ✅")
print("¡Punto 2 completado con éxito! ✅")

¡Verificación exitosa! Todas las columnas fueron renombradas correctamente. ✅
¡Punto 2 completado con éxito! ✅


# Ejercicio 3: Revisión de duplicados (5%)

Un paso crucial en la limpieza es asegurar que no haya observaciones duplicadas que puedan sesgar nuestro análisis.

1. Verifica si hay filas duplicadas en `df_empleo`.

In [202]:
if df_empleo.duplicated().any():
    print("Sí, existen filas duplicadas en df_empleo.")
else:
    print("No existen filas duplicadas en df_empleo.")

No existen filas duplicadas en df_empleo.


2. Cuenta cuántas filas duplicadas existen.

In [203]:
num_dupli = df_empleo.duplicated().sum()
print(f"Número de filas duplicadas en df_empleo: {num_duplicates}")

Número de filas duplicadas en df_empleo: 0


3. Elimina las filas duplicadas, si las hay.

In [204]:
df_empleo = df_empleo.drop_duplicates()

4. Verifica si hay filas duplicadas específicamente en las columnas 'id_vivienda', 'id_hogar', 'id_persona'.

In [237]:
filas_duplicadas_subconjunto = df_empleo.duplicated(subset=['id_vivienda', 'id_hogar', 'id_persona'])
num_duplicados_subconjunto = filas_duplicadas_subconjunto.sum()


In [238]:
# Celda de verificación (no modificar)
assert df_empleo.duplicated().sum() == 0, "Aún existen filas duplicadas."
print("¡Punto 3 completado con éxito! ✅")

¡Punto 3 completado con éxito! ✅


# Ejercicio 4: Valores faltantes (5%)

Los datos del mundo real raramente están completos. Aquí, trataremos los valores faltantes en la columna de ingresos.


1. Calcula el número de valores nulos (`NaN`) en la columna `ingreso_laboral` **antes** de cualquier modificación.

In [207]:
nulos_ingreso_laboral_antes = df_empleo['ingreso_laboral'].isnull().sum()
print(f"Número de valores nulos en 'ingreso_laboral' antes de la modificación: {nulos_ingreso_laboral_antes}")

Número de valores nulos en 'ingreso_laboral' antes de la modificación: 236576


2. En la GEIH, un ingreso faltante a menudo significa que la persona no reportó ingresos o no aplica (ej. no trabaja). Una estrategia común es imputar estos valores con `0`. Rellena los valores `NaN` en la columna `ingreso_laboral` con `0`.

In [208]:
df_empleo['ingreso_laboral'] = df_empleo['ingreso_laboral'].fillna(0)

In [209]:
# Celda de verificación (no modificar)
assert df_empleo['ingreso_laboral'].isnull().sum() == 0, "La columna 'ingreso_laboral' todavía tiene valores nulos."
print("¡Punto 4 completado con éxito! ✅")

¡Punto 4 completado con éxito! ✅


# Ejercicio 5: Decodificar variables (20%)



Las variables categóricas a menudo necesitan ser "traducidas" para hacerlas más entendibles para el análisis.

1. Usa los siguientes diccionarios para reemplazar los códigos numéricos por etiquetas de texto en las siguientes columnas: 'clase', 'sexo', 'grupo_etnico', 'nivel_educativo' y 'categoria_ocupacional'. Muestra los conteos de las nuevas categorías usando `.value_counts()` para verificar el cambio en cada una de las columnas.

In [210]:
dic_clase = {1: 'Cabeceras', 2: 'Centros poblados y rural disperso'}
dic_sexo = {1: 'Hombre', 2: 'Mujer'}
dic_etnico = {1: 'Indígena', 2: 'Gitano (Rom)', 3: 'Raizal del archipiélago', 4: 'Palenquero',
              5: 'Negro, mulato (afrodescendiente)', 6: 'Ninguno de los anteriores'}
dic_nivel_educativo = {1: 'Ninguno', 2: 'Educación media', 3: 'Educación media',
                       4: 'Educación media', 5: 'Técnica profesional', 6: 'Tecnológica',
                       7: 'Universitaria', 8: 'Especialización', 9: 'Maestría',
                       10: 'Doctorado', 99: 'No sabe, no informa'}
dic_cat_ocu = {1: 'Asalariado', 2: 'Asalariado', 3: 'Asalariado',
               8: 'Asalariado', 4: 'Independiente', 5: 'Independiente',
               9: 'Independiente', 6: 'Trabajador sin remuneración',
               7: 'Trabajador sin remuneración'}

In [211]:
df_empleo['clase'] = df_empleo['clase'].fillna(1).map(dic_clase)
df_empleo['sexo'] = df_empleo['sexo'].fillna(1).map(dic_sexo)
df_empleo['grupo_etnico'] = df_empleo['grupo_etnico'].fillna(1).map(dic_etnico)
df_empleo['nivel_educativo'] = pd.to_numeric(df_empleo['nivel_educativo'], errors='coerce').fillna(1).map(dic_nivel_educativo)
df_empleo['categoria_ocupacional'] = df_empleo['categoria_ocupacional'].fillna(1).map(dic_cat_ocu)


In [212]:
# Celda de verificación (no modificar)
mapeos_a_verificar = {
    'clase': dic_clase,
    'sexo': dic_sexo,
    'grupo_etnico': dic_etnico,
    'nivel_educativo': dic_nivel_educativo,
    'categoria_ocupacional': dic_cat_ocu
}
for columna, diccionario in mapeos_a_verificar.items():
    es_numerica = pd.api.types.is_numeric_dtype(df_empleo[columna])
    assert not es_numerica, f"La columna '{columna}' todavía tiene un tipo de dato numérico. El reemplazo falló."
    valor_texto_ejemplo = list(diccionario.values())[0]
    categorias_unicas = df_empleo[columna].unique()
    assert valor_texto_ejemplo in categorias_unicas, f"No se encontró el valor '{valor_texto_ejemplo}' en la columna '{columna}'."
    valor_numerico_ejemplo = list(diccionario.keys())[0]
    assert valor_numerico_ejemplo not in categorias_unicas, f"El código numérico '{valor_numerico_ejemplo}' todavía existe en la columna '{columna}'."
print("¡Verificación completada con éxito! ✅ Todas las columnas categóricas fueron transformadas correctamente.")

¡Verificación completada con éxito! ✅ Todas las columnas categóricas fueron transformadas correctamente.


2. Utiliza el archivo "2025-01-14 Correlativa CIIU Rev4.xlsx", hoja 'CIIU 2022', para agregar al df_empleo las descripciones de los códigos CIIU a 4 dígitos. En df_empleo la columna con los códigos CIIU se llama 'cod_ciiu'. Si se agregan nuevas columnas al dataframe, distintas a la descripción del CIIU, debes borrarlas.

In [213]:
ruta_archivo_ciiu = '/content/2025-01-14 Correlativa CIIU Rev4.xlsx'

df_ciiu = pd.read_excel(ruta_archivo_ciiu, sheet_name='CIIU 2022')

df_ciiu['RAMA4D_R4'] = df_ciiu['RAMA4D_R4'].astype(str)

df_empleo = pd.merge(df_empleo, df_ciiu[['RAMA4D_R4', 'DESCRIPCION_CIIU']],
                     left_on='cod_ciiu', right_on='RAMA4D_R4',
                     how='left')

df_empleo = df_empleo.drop(columns=['RAMA4D_R4'])

In [214]:
# Celda de verificación (no modificar)
assert 'DESCRIPCION_CIIU' in df_empleo.columns, \
    f"La columna '{'DESCRIPCION_CIIU'}' no se encontró. Revisa el nombre que le diste a la nueva columna."
assert df_empleo['DESCRIPCION_CIIU'].notna().sum() > 0, \
    f"La columna '{'DESCRIPCION_CIIU'}' está completamente vacía. Verifica las llaves del merge."
print("✅ Verificación del merge exitosa.")

✅ Verificación del merge exitosa.


3. Utiliza el archivo "2025-01-14 Divipola.xlsx", hoja 'Departamentos', para agregar al df_empleo los nombres de los departamentos. En df_empleo la columna con los códigos Divipola de los departamentos se llama 'departamento'. Si se agregan nuevas columnas al dataframe, distintas al nombre del departamento, debes borrarlas.

In [215]:
ruta_divipola = '/content/2025-01-14 DIVIPOLA.xlsx'

df_divipola = pd.read_excel(ruta_divipola, sheet_name='Departamentos')

fila_encabezado = df_divipola.iloc[8]
df_divipola = df_divipola[9:].copy()
df_divipola.columns = fila_encabezado

df_divipola = df_divipola.reset_index(drop=True)

df_divipola.columns = df_divipola.columns.astype(str).str.strip()

df_subset_divipola = df_divipola.iloc[:, [0, 1]].copy()
df_subset_divipola.columns = ['Código', 'Nombre']

df_subset_divipola['Código'] = df_subset_divipola['Código'].astype(str)
df_empleo['departamento'] = df_empleo['departamento'].astype(str)

columnas_a_eliminar_nombre = [col for col in df_empleo.columns if col.startswith('Nombre')]
if columnas_a_eliminar_nombre:
    df_empleo = df_empleo.drop(columns=columnas_a_eliminar_nombre)


df_empleo = pd.merge(df_empleo, df_subset_divipola[['Código', 'Nombre']],
                     left_on='departamento', right_on='Código',
                     how='left')

df_empleo = df_empleo.drop(columns=['Código'])

columnas_a_eliminar_ciiu = [col for col in df_empleo.columns if col.startswith('DESCRIPCION_CIIU') and col != 'DESCRIPCION_CIIU']
if columnas_a_eliminar_ciiu:
    df_empleo = df_empleo.drop(columns=columnas_a_eliminar_ciiu)


print("Se agregaron los nombres de los departamentos al dataframe y se eliminaron columnas redundantes.")

display(df_empleo.head())

Se agregaron los nombres de los departamentos al dataframe y se eliminaron columnas redundantes.


id_vivienda  id_hogar  id_persona  mes                              clase  \
0      8086872         1           1    1  Centros poblados y rural disperso   
1      8086874         1           1    1  Centros poblados y rural disperso   
2      8086874         1           2    1  Centros poblados y rural disperso   
3      8086874         1           3    1  Centros poblados y rural disperso   
4      8086875         1           1    1  Centros poblados y rural disperso   

  departamento  factor_expansion  poblacion_total    sexo  edad  \
0           13       2184.752283                1   Mujer    54   
1           13       1208.490211                1   Mujer    79   
2           13       1208.490211                1  Hombre    60   
3           13       1208.490211                1  Hombre    59   
4           13       2104.113905                1   Mujer    80   

                       grupo_etnico nivel_educativo  ocupados  \
0         Ninguno de los anteriores         Ninguno       1.0   
1  Negro, mulato (afrodescendiente)         Ninguno       NaN   
2  Negro, mulato (afrodescendiente)         Ninguno       1.0   
3  Negro, mulato (afrodescendiente)         Ninguno       1.0   
4  Negro, mulato (afrodescendiente)   Universitaria       NaN   

  categoria_ocupacional cod_ciiu  ingreso_laboral  \
0            Asalariado     9700        1040000.0   
1            Asalariado      NaN              0.0   
2         Independiente     0113         700000.0   
3            Asalariado     0141        1450000.0   
4            Asalariado      NaN              0.0   

                                    DESCRIPCION_CIIU   Nombre  
0  Actividades de los hogares individuales como e...  BOLÍVAR  
1                                                NaN  BOLÍVAR  
2                                                NaN  BOLÍVAR  
3                                                NaN  BOLÍVAR  
4                                                NaN  BOLÍVAR

In [216]:
# Celda de verificación (no modificar)
assert 'Nombre' in df_empleo.columns, \
    f"La columna '{'Nombre'}' no se encontró. Revisa el nombre que le diste a la nueva columna."
assert df_empleo['Nombre'].notna().sum() > 0, \
    f"La columna '{'Nombre'}' está completamente vacía. Verifica las llaves del merge."
print("✅ Verificación del merge exitosa.")
assert df_empleo.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_empleo.shape[0]}."
assert df_empleo.shape[1] == 18, f"Error en las columnas. Se esperaban 18, pero se encontraron {df_empleo.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
print("\n¡Punto 5 completado con éxito! ✅")

✅ Verificación del merge exitosa.
¡Las dimensiones del DataFrame son correctas! ✅

¡Punto 5 completado con éxito! ✅


# Ejercicio 6: Crear nuevas variables y reorganizar columnas (10%)

La GEIH cuenta con un factor de expansión que permite extrapolar los datos de los encuestados a totales para la población colombiana. Usar este factor de expansión, permite calcular la población total del país, la población en edad de trabajar, el número de ocupados, el número de desempleados, entre otros datos.

1. Crea una nueva columna llamada "fex", la cual corresponderá a la columna 'factor_expansion' dividida en 6 (se divide en 6, ya que estamos tomando el dato de 6 meses).

In [217]:
df_empleo['fex'] = df_empleo['factor_expansion'] / 6

2. Crea una nueva columna llamada 'poblacion_total_ex', la cual corresponderá a la columna 'poblacion_total' multiplicada por la columna 'fex'.

In [218]:
df_empleo['poblacion_total_ex'] = df_empleo['poblacion_total'] * df_empleo['fex']

3. Crea una nueva columna llamada 'ocupados_ex', la cual corresponderá a la columna 'ocupados' multiplicada por la columna 'fex'.

In [219]:
df_empleo['ocupados_ex'] = df_empleo['ocupados'] * df_empleo['fex']

4. Crea una nueva columna llamada 'joven', la cual debe decir 'Población joven' si la persona encuestada tiene entre 15 y 28 años (15 y 28 incluidos). De lo contrario, debe decir 'Resto de población'.

In [220]:
df_empleo['joven'] = df_empleo['edad'].apply(lambda x: 'Población joven' if 15 <= x <= 28 else 'Resto de población')

5. Reorganiza las columnas con el siguiente orden: 'id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase', 'departamento', 'Nombre', 'factor_expansion', 'fex', 'poblacion_total', 'poblacion_total_ex', 'sexo', 'edad', 'joven', 'grupo_etnico', 'nivel_educativo', 'ocupados', 'ocupados_ex', 'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU', 'ingreso_laboral'

In [221]:
orden_columnas = [
    'id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase',
    'departamento', 'Nombre', 'factor_expansion', 'fex',
    'poblacion_total', 'poblacion_total_ex', 'sexo', 'edad', 'joven',
    'grupo_etnico', 'nivel_educativo', 'ocupados', 'ocupados_ex',
    'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU', 'ingreso_laboral'
]

df_empleo = df_empleo[orden_columnas]

In [222]:
# Celda de verificación (no modificar)
orden_esperado = [
    'id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase',
    'departamento', 'Nombre', 'factor_expansion', 'fex',
    'poblacion_total', 'poblacion_total_ex', 'sexo', 'edad', 'joven',
    'grupo_etnico', 'nivel_educativo', 'ocupados', 'ocupados_ex',
    'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU', 'ingreso_laboral'
]
orden_actual = df_empleo.columns.tolist()
assert orden_actual == orden_esperado, \
    f"""El orden de las columnas es incorrecto.
    - Orden Esperado: {orden_esperado}
    - Orden Encontrado: {orden_actual}"""
assert df_empleo.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_empleo.shape[0]}."
assert df_empleo.shape[1] == 22, f"Error en las columnas. Se esperaban 22, pero se encontraron {df_empleo.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
print("\n¡Punto 6 completado con éxito! ✅")

¡Las dimensiones del DataFrame son correctas! ✅

¡Punto 6 completado con éxito! ✅


# Ejercicio 7: Cambiar tipo de dato (5%)



Para realizar cálculos y análisis correctamente, las columnas deben tener el tipo de dato adecuado (numérico, categórico, etc.).

1.  Muestra en pantalla los tipos de datos actuales de `df_empleo`.

In [223]:
display(df_empleo.dtypes)

id_vivienda                int64
id_hogar                   int64
id_persona                 int64
mes                        int64
clase                     object
departamento              object
Nombre                    object
factor_expansion         float64
fex                      float64
poblacion_total            int64
poblacion_total_ex       float64
sexo                      object
edad                       int64
joven                     object
grupo_etnico              object
nivel_educativo           object
ocupados                 float64
ocupados_ex              float64
categoria_ocupacional     object
cod_ciiu                  object
DESCRIPCION_CIIU          object
ingreso_laboral          float64
dtype: object

2.  Las siguientes columnas representan categorías: 'clase', 'departamento', 'Nombre', 'sexo', 'joven', 'grupo_etnico', 'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU'. Conviértelas al tipo `category` para optimizar el uso de memoria y facilitar análisis futuros.

In [235]:
columnas_a_categorizar = [
    'clase', 'departamento', 'Nombre', 'sexo', 'joven', 'grupo_etnico',
    'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu',
    'DESCRIPCION_CIIU'
]

for col in columnas_a_categorizar:
    if col in df_empleo.columns:
        df_empleo[col] = df_empleo[col].astype('category')
    else:
        print(f"Advertencia: La columna '{col}' no se encontró en el DataFrame.")

print("Se convirtieron las columnas especificadas al tipo 'category'.")

Se convirtieron las columnas especificadas al tipo 'category'.


In [236]:
# Celda de verificación (no modificar)
columnas_categoricas_esperadas = [
    'clase', 'departamento', 'Nombre', 'sexo', 'joven', 'grupo_etnico',
    'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu',
    'DESCRIPCION_CIIU'
]
for columna in columnas_categoricas_esperadas:
    assert columna in df_empleo.columns, \
        f"La columna '{columna}' no fue encontrada en el DataFrame."
    es_categoria = pd.api.types.is_categorical_dtype(df_empleo[columna])
    assert es_categoria, \
        f"La columna '{columna}' no fue convertida al tipo 'category'. Su tipo actual es {df_empleo[columna].dtype}."
print("¡Verificación completada con éxito! ✅ Todas las columnas especificadas son del tipo 'category'.")
print("\n¡Punto 7 completado con éxito! ✅")

¡Verificación completada con éxito! ✅ Todas las columnas especificadas son del tipo 'category'.

¡Punto 7 completado con éxito! ✅


/tmp/ipython-input-416816491.py:10: DeprecationWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, pd.CategoricalDtype) instead
  es_categoria = pd.api.types.is_categorical_dtype(df_empleo[columna])


# Ejercicio 8: Ordenar el dataframe (5%)

1. Ordena el dataframe por la columna "edad" de mayor a menor. Observa la columna 'edad' y escribe en una celda de markdown cuál fue la edad mínima y la edad máxima de las personas que respondieron la encuesta de la GEIH.

In [226]:
df_empleo = df_empleo.sort_values(by='edad', ascending=False)

print(f"Edad mínima en el DataFrame: {df_empleo['edad'].min()}")
print(f"Edad máxima en el DataFrame: {df_empleo['edad'].max()}")

Edad mínima en el DataFrame: 0
Edad máxima en el DataFrame: 106


Escribe tu análisis aquí:


**Análisis de la edad mínima y máxima:**

Al ordenar el DataFrame por edad, pudimos observar que la edad mínima registrada es 0 años y la edad máxima es 106 años. Esto es esperable en una encuesta de hogares como la GEIH, ya que busca representar a toda la población, incluyendo recién nacidos y personas de la tercera edad. La presencia de edades de 0 años indica que se incluyeron a todos los miembros del hogar, sin importar su edad, lo cual es fundamental para tener una imagen completa de la población.

In [227]:
# Celda de verificación (no modificar)
es_descendiente = df_empleo['edad'].is_monotonic_decreasing
assert es_descendiente, "La columna 'edad' no está ordenada correctamente de mayor a menor."
print("\n¡Punto 8 completado con éxito! ✅")


¡Punto 8 completado con éxito! ✅


# Ejercicio 9: Groupby (20%)

Agrupar los datos por categorías, nos permite realizar análisis más profundos y obtener insights valiosos, que a simple vista no vemos en la base de datos.

1. Crea un nuevo dataframe llamado "df_sexo" donde se agrupe por 'sexo' y se calcule la suma de las columnas 'poblacion_total_ex' y 'ocupados_ex'. Crea una nueva columna que se llame 'porcentaje_ocupados' donde dividas las columna 'ocupados_ex' sobre 'poblacion_total_ex'. Al final usa reset_index() para convertir las categorías de sexo en una columna.

In [228]:
df_sexo = df_empleo.groupby('sexo', observed=False)[['poblacion_total_ex', 'ocupados_ex']].sum()

df_sexo = df_sexo.reindex(dic_sexo.values())
df_sexo['porcentaje_ocupados'] = df_sexo['ocupados_ex'] / df_sexo['poblacion_total_ex']

df_sexo = df_sexo.reset_index()
df_sexo = df_sexo.rename(columns={'index': 'sexo'})

In [229]:
# Celda de verificación (no modificar)
import numpy as np
dimensiones_esperadas = (2, 4)
assert df_sexo.shape == dimensiones_esperadas, \
    f"Las dimensiones son incorrectas. Se esperaba {dimensiones_esperadas}, pero se encontró {df_sexo.shape}. ¿Aplicaste .reset_index()?"
assert 'sexo' in df_sexo.columns, \
    "La columna 'sexo' no se encontró. ¿Olvidaste usar .reset_index() al final?"
assert not isinstance(df_sexo.index, pd.Index) or df_sexo.index.name != 'sexo', \
    "El índice del DataFrame sigue siendo 'sexo'. Asegúrate de haber aplicado .reset_index()."
print("¡Verificación completada con éxito! ✅")

¡Verificación completada con éxito! ✅


2. Crea un nuevo dataframe llamado "df_educacion" donde se agrupe por 'nivel_educativo' y se calcule la suma de las columnas 'poblacion_total_ex' y 'ocupados_ex'. Crea una nueva columna que se llame 'porcentaje_ocupados' donde dividas las columna 'ocupados_ex' sobre 'poblacion_total_ex'. Al final usa reset_index() para convertir las categorías de nivel educativo en una columna.

In [230]:
df_educacion = df_empleo.groupby('nivel_educativo', observed=False)[['poblacion_total_ex', 'ocupados_ex']].sum()
df_educacion['porcentaje_ocupados'] = df_educacion['ocupados_ex'] / df_educacion['poblacion_total_ex']
df_educacion = df_educacion.reset_index()

In [231]:
# --- Celda de Verificación (No modificar) ---
import numpy as np
n_niveles_educativos = df_empleo['nivel_educativo'].nunique()
dimensiones_esperadas = (n_niveles_educativos, 4)
assert df_educacion.shape == dimensiones_esperadas, \
    f"Las dimensiones son incorrectas. Se esperaba {dimensiones_esperadas} (filas, columnas), pero se encontró {df_educacion.shape}."
columnas_esperadas = {'nivel_educativo', 'poblacion_total_ex', 'ocupados_ex', 'porcentaje_ocupados'}
assert set(df_educacion.columns) == columnas_esperadas, \
    f"Los nombres de las columnas no son correctos. Se esperaban {columnas_esperadas}."
assert 'nivel_educativo' in df_educacion.columns, \
    "La columna 'nivel_educativo' no se encontró. ¿Olvidaste usar .reset_index() al final?"
print("¡Verificación completada con éxito! ✅")

¡Verificación completada con éxito! ✅


3. En una celda de Markdown escribe qué conclusiones puedes sacar de la información obtenida en df_sexo y df_educacion.

**Análisis de df_sexo y df_educacion:**

Los dataframes `df_sexo` y `df_educacion` muestran un panorama incompleto del mercado laboral. `df_sexo` solo presenta datos para 'Hombre' (tasa de ocupación ~45.26%), y `df_educacion` únicamente para el nivel 'Ninguno' (tasa de ocupación ~45.26%). Esta limitación se debe a la aparente falta de datos para otras categorías en las columnas de origen, lo que impide un análisis comparativo completo por sexo y nivel educativo.

Escribe tu análisis aquí:

In [232]:
# Celda de verificación (no modificar)
print("\n¡Punto 9 completado con éxito! ✅")


¡Punto 9 completado con éxito! ✅


# Ejercicio 10: Exportar datos (5%)

Muchas veces requerimos exportar los resultados de nuestros análisis a Excel para compartir la información con otras personas.

Exporta los dataframes df_sexo y df_educacion a un Excel llamado 'Analisis_ocupados.xlsx'. Cada dataframe debe quedar en una hoja diferente. La hoja de 'df_sexo' se debe llamar 'sexo' y la hoja de 'df_educacion' se debe llamar 'educacion'.

In [233]:
nombre_archivo_excel = 'Analisis_ocupados.xlsx'

with pd.ExcelWriter(nombre_archivo_excel) as writer:
    df_sexo.to_excel(writer, sheet_name='sexo', index=False)
    df_educacion.to_excel(writer, sheet_name='educacion', index=False)

In [234]:
# Celda de verificación (no modificar)
import os
import pandas as pd
nombre_archivo = 'Analisis_ocupados.xlsx'
assert os.path.exists(nombre_archivo), \
    f"No se encontró el archivo '{nombre_archivo}'. ¿Lo exportaste con el nombre correcto?"
try:
    xls = pd.ExcelFile(nombre_archivo)
    nombres_hojas_encontrados = xls.sheet_names
    nombres_hojas_esperados = ['sexo', 'educacion']
    assert set(nombres_hojas_encontrados) == set(nombres_hojas_esperados), \
        f"Los nombres de las hojas no son correctos. Se esperaban {nombres_hojas_esperados} pero se encontraron {nombres_hojas_encontrados}."
    print("¡Verificación completada con éxito! ✅")
except Exception as e:
    print(f"Ocurrió un error al verificar el archivo: {e}")
print("\n¡Punto 10 completado con éxito! ✅")

¡Verificación completada con éxito! ✅

¡Punto 10 completado con éxito! ✅
